http-server-netty/src/main/java/io/micronaut/http/server/netty/HttpContentProcessorAsReactiveProcessor.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2017-2022 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.server.netty;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.http.netty.stream.StreamedHttpMessage;
+import io.netty.buffer.ByteBufHolder;
+import org.reactivestreams.Processor;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.Queue;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Extension of {@link HttpContentProcessorAsReactiveSubscriber} that is also a {@link Publisher}.
+ * This class is a bit more complicated because it has to deal with downstream demand, possibly
+ * concurrently.
+ *
+ * @since 4.0.0
+ * @author Jonas Konrad
+ */
+@Internal
+public final class HttpContentProcessorAsReactiveProcessor extends HttpContentProcessorAsReactiveSubscriber implements Processor<ByteBufHolder, Object> {
+    private static final Object END = new Object();
```

Comment:
it won't work unfortunately, `.doOnRequest(demand -> subscription.request(demand));` is incorrect. since a processor may yield multiple items, downstream demand does not have to equal upstream demand

Sinks.Many does not expose api to check whether upstream demand is there, maybe could use a custom queue. but then concurrency becomes an issue since we cant use the reactor internal concurrency structure.

this could literally be replaced with a flatMap though, but it would also need completion/error handling, so it would have a bunch of nested publishers.

Reply:
this is very similar to SpreadProcessor in serde btw

Reply:
It should work correctly, the sink is not going to push more items it’s just going buffer them.

Reply:
Yeah, good point. But do we respect the requesting from the Netty content publisher?

Reply:
yea it translates to netty read calls which generates backpressure

Reply:
Maybe you can use Flux.using? I’m a bit confused why do we need to call complete…

Anyway it might be a bit to reactive but might save us from potential bugs, we would need to profile some basic cases.

Reply:
complete is used e.g. by the xml processor, it accumulates all data and only fires it downstream on the call to complete.

i dont think `using` will help.

maybe the perf hit is not so bad. this processor mostly exists for places where reactive is unavoidable, e.g. when the user has a publisher parameter anyway. in the long term we should strive to not use reactive when the user doesnt want reactive anyway, so this wouldnt be hit.

im going to switch to the flatmap approach for now, we can always dig up the old code if it becomes perf critical. i dont want to microbenchmark them against each other right now.